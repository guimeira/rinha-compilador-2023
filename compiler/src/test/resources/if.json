{"name":"if.rinha","expression":{"kind":"Let","name":{"text":"a","location":{"start":4,"end":5,"filename":"if.rinha"}},"value":{"kind":"Int","value":10,"location":{"start":8,"end":10,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"b","location":{"start":16,"end":17,"filename":"if.rinha"}},"value":{"kind":"Int","value":20,"location":{"start":20,"end":22,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"c","location":{"start":28,"end":29,"filename":"if.rinha"}},"value":{"kind":"If","condition":{"kind":"Binary","lhs":{"kind":"Var","text":"a","location":{"start":35,"end":36,"filename":"if.rinha"}},"op":"Gt","rhs":{"kind":"Var","text":"b","location":{"start":39,"end":40,"filename":"if.rinha"}},"location":{"start":35,"end":40,"filename":"if.rinha"}},"then":{"kind":"Binary","lhs":{"kind":"Int","value":10,"location":{"start":44,"end":46,"filename":"if.rinha"}},"op":"Mul","rhs":{"kind":"Var","text":"a","location":{"start":47,"end":48,"filename":"if.rinha"}},"location":{"start":44,"end":48,"filename":"if.rinha"}},"otherwise":{"kind":"Binary","lhs":{"kind":"Int","value":10,"location":{"start":58,"end":60,"filename":"if.rinha"}},"op":"Mul","rhs":{"kind":"Var","text":"b","location":{"start":61,"end":62,"filename":"if.rinha"}},"location":{"start":58,"end":62,"filename":"if.rinha"}},"location":{"start":32,"end":64,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":70,"end":71,"filename":"if.rinha"}},"value":{"kind":"Print","value":{"kind":"Var","text":"c","location":{"start":80,"end":81,"filename":"if.rinha"}},"location":{"start":74,"end":82,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"d","location":{"start":88,"end":89,"filename":"if.rinha"}},"value":{"kind":"If","condition":{"kind":"Binary","lhs":{"kind":"Var","text":"b","location":{"start":95,"end":96,"filename":"if.rinha"}},"op":"Gt","rhs":{"kind":"Var","text":"a","location":{"start":99,"end":100,"filename":"if.rinha"}},"location":{"start":95,"end":100,"filename":"if.rinha"}},"then":{"kind":"Binary","lhs":{"kind":"Int","value":10,"location":{"start":104,"end":106,"filename":"if.rinha"}},"op":"Mul","rhs":{"kind":"Var","text":"a","location":{"start":107,"end":108,"filename":"if.rinha"}},"location":{"start":104,"end":108,"filename":"if.rinha"}},"otherwise":{"kind":"Binary","lhs":{"kind":"Int","value":10,"location":{"start":118,"end":120,"filename":"if.rinha"}},"op":"Mul","rhs":{"kind":"Var","text":"b","location":{"start":121,"end":122,"filename":"if.rinha"}},"location":{"start":118,"end":122,"filename":"if.rinha"}},"location":{"start":92,"end":124,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":130,"end":131,"filename":"if.rinha"}},"value":{"kind":"Print","value":{"kind":"Var","text":"d","location":{"start":140,"end":141,"filename":"if.rinha"}},"location":{"start":134,"end":142,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"f","location":{"start":148,"end":149,"filename":"if.rinha"}},"value":{"kind":"Function","parameters":[{"text":"b1","location":{"start":155,"end":157,"filename":"if.rinha"}},{"text":"b2","location":{"start":159,"end":161,"filename":"if.rinha"}}],"value":{"kind":"If","condition":{"kind":"Var","text":"b1","location":{"start":173,"end":175,"filename":"if.rinha"}},"then":{"kind":"If","condition":{"kind":"Var","text":"b2","location":{"start":186,"end":188,"filename":"if.rinha"}},"then":{"kind":"Str","value":"b1b2","location":{"start":198,"end":204,"filename":"if.rinha"}},"otherwise":{"kind":"Str","value":"b1!b2","location":{"start":224,"end":231,"filename":"if.rinha"}},"location":{"start":183,"end":237,"filename":"if.rinha"}},"otherwise":{"kind":"If","condition":{"kind":"Var","text":"b2","location":{"start":256,"end":258,"filename":"if.rinha"}},"then":{"kind":"Str","value":"!b1b2","location":{"start":268,"end":275,"filename":"if.rinha"}},"otherwise":{"kind":"Str","value":"!b1!b2","location":{"start":295,"end":303,"filename":"if.rinha"}},"location":{"start":253,"end":309,"filename":"if.rinha"}},"location":{"start":170,"end":313,"filename":"if.rinha"}},"location":{"start":152,"end":315,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":321,"end":322,"filename":"if.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"f","location":{"start":331,"end":332,"filename":"if.rinha"}},"arguments":[{"kind":"Bool","value":true,"location":{"start":333,"end":337,"filename":"if.rinha"}},{"kind":"Bool","value":true,"location":{"start":338,"end":342,"filename":"if.rinha"}}],"location":{"start":331,"end":343,"filename":"if.rinha"}},"location":{"start":325,"end":344,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":350,"end":351,"filename":"if.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"f","location":{"start":360,"end":361,"filename":"if.rinha"}},"arguments":[{"kind":"Bool","value":true,"location":{"start":362,"end":366,"filename":"if.rinha"}},{"kind":"Bool","value":false,"location":{"start":367,"end":372,"filename":"if.rinha"}}],"location":{"start":360,"end":373,"filename":"if.rinha"}},"location":{"start":354,"end":374,"filename":"if.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":380,"end":381,"filename":"if.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"f","location":{"start":390,"end":391,"filename":"if.rinha"}},"arguments":[{"kind":"Bool","value":false,"location":{"start":392,"end":397,"filename":"if.rinha"}},{"kind":"Bool","value":true,"location":{"start":398,"end":402,"filename":"if.rinha"}}],"location":{"start":390,"end":403,"filename":"if.rinha"}},"location":{"start":384,"end":404,"filename":"if.rinha"}},"next":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"f","location":{"start":412,"end":413,"filename":"if.rinha"}},"arguments":[{"kind":"Bool","value":false,"location":{"start":414,"end":419,"filename":"if.rinha"}},{"kind":"Bool","value":false,"location":{"start":420,"end":425,"filename":"if.rinha"}}],"location":{"start":412,"end":426,"filename":"if.rinha"}},"location":{"start":406,"end":427,"filename":"if.rinha"}},"location":{"start":376,"end":427,"filename":"if.rinha"}},"location":{"start":346,"end":427,"filename":"if.rinha"}},"location":{"start":317,"end":427,"filename":"if.rinha"}},"location":{"start":144,"end":427,"filename":"if.rinha"}},"location":{"start":126,"end":427,"filename":"if.rinha"}},"location":{"start":84,"end":427,"filename":"if.rinha"}},"location":{"start":66,"end":427,"filename":"if.rinha"}},"location":{"start":24,"end":427,"filename":"if.rinha"}},"location":{"start":12,"end":427,"filename":"if.rinha"}},"location":{"start":0,"end":427,"filename":"if.rinha"}},"location":{"start":0,"end":427,"filename":"if.rinha"}}
