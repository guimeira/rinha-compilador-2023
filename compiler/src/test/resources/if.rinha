let a = 10;
let b = 20;
let c = if(a > b) { 10*a } else { 10*b };
let _ = print(c);
let d = if(b > a) { 10*a } else { 10*b };
let _ = print(d);
let f = fn(b1, b2) => {
  if(b1) {
    if(b2) {
      "b1b2"
    } else {
      "b1!b2"
    }
  } else {
    if(b2) {
      "!b1b2"
    } else {
      "!b1!b2"
    }
  }
};
let _ = print(f(true,true));
let _ = print(f(true,false));
let _ = print(f(false,true));
print(f(false,false))
