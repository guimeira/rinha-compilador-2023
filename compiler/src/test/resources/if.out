200
100
b1b2
b1!b2
!b1b2
!b1!b2

